### DISCUSSION TITLE
[assemble] Jlink fails due to missing JAR's

### DISCUSSION THREAD

**Discussion by stackoverflow2** [jreleaser/jreleaser#848] [Q&A]
I am trying to use Jreleaser: assemble  plugin to create distribution for my quarkus app and target JDK version is 11. But assembler is failing with below error : 

`[ERROR]   [jlink] Exception in thread "main" java.lang.module.FindException: Module java.xml.bind not found, required by java.ws.rs
[ERROR]   [jlink] 	at java.base/java.lang.module.Resolver.findFail(Resolver.java:877)
[ERROR]   [jlink] 	at java.base/java.lang.module.Resolver.resolve(Resolver.java:191)
[ERROR]   [jlink] 	at java.base/java.lang.module.Resolver.resolve(Resolver.java:140)
[ERROR]   [jlink] 	at java.base/java.lang.module.Configuration.resolve(Configuration.java:422)
[ERROR]   [jlink] 	at java.base/java.lang.module.Configuration.resolve(Configuration.java:256)
[ERROR]   [jlink] 	at jdk.jdeps/com.sun.tools.jdeps.JdepsConfiguration$Builder.build(JdepsConfiguration.java:564)
[ERROR]   [jlink] 	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.buildConfig(JdepsTask.java:603)
[ERROR]   [jlink] 	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.run(JdepsTask.java:557)
[ERROR]   [jlink] 	at jdk.jdeps/com.sun.tools.jdeps.JdepsTask.run(JdepsTask.java:533)
[ERROR]   [jlink] 	at jdk.jdeps/com.sun.tools.jdeps.Main.main(Main.java:49)
[ERROR]   [jlink] 
[INFO]  Writing output properties to main\target\jreleaser\output.properties
[ERROR] JReleaser failed after 9.837 s
org.jreleaser.util.JReleaserException: Unexpected error`

**Comment by aalmiray** [comment]
Could you share details on your configuration? Be aware of https://github.com/jreleaser/jreleaser/issues/844 The current workaround is to use relative paths instead of absolute paths.

**Reply by stackoverflow2** [reply]
Yes, already noted https://github.com/jreleaser/jreleaser/issues/844 and using relative paths.

Please find the configuration : 

[INFO] Detecting the operating system and CPU architecture
[INFO] ------------------------------------------------------------------------
[INFO] os.detected.name: windows
[INFO] os.detected.arch: x86_64
[INFO] os.detected.bitness: 64
[INFO] os.detected.version: 10.0
[INFO] os.detected.version.major: 10
[INFO] os.detected.version.minor: 0
[INFO] os.detected.classifier: windows-x86_64

```xml
<plugin>
          <groupId>org.jreleaser</groupId>
          <artifactId>jreleaser-maven-plugin</artifactId>
          <version>0.9.0</version>
          <configuration>
            <gitRootSearch>true</gitRootSearch>
            <jreleaser>
              <environment>
                <properties>
                  <jdkFilePrefix>zulu11.50.19-ca-jdk11.0.12</jdkFilePrefix>
                </properties>
              </environment>
              <assemble>
                <jlink>
                  <comet-aq-webservice>
                    <active>ALWAYS</active>
                    <java>
                      <mainClass>io.quarkus.bootstrap.runner.QuarkusEntryPoint</mainClass>
                    </java>
                    <imageName>{{distributionName}}-{{projectVersion}}</imageName>
                    <additionalModuleNames>
                      <additionalModuleName>java.xml.bind</additionalModuleName>
                    </additionalModuleNames>
                    <targetJdks>
                      <targetJdk>
                        <path>main/target/jdks/zulu11Windows/{{jdkFilePrefix}}-win_x64</path>
                        <platform>windows-x86_64</platform>
                      </targetJdk>
                    </targetJdks>
                    <copyJars>true</copyJars>
                    <jdeps>
                      <multiRelease>11</multiRelease>
                      <ignoreMissingDeps>true</ignoreMissingDeps>
                    </jdeps>
                    <args>
                      <arg>--bind-services</arg>
                    </args>
                    <mainJar>
                      <path>main/target/quarkus-app/quarkus-run.jar</path>
                    </mainJar>
                    <jars>
                      <jar>
                        <pattern>main/target/quarkus-app/lib/**/*.jar</pattern>
                      </jar>
                    </jars>
                    <fileSets>
                      <fileSet>
                        <input>main/target/quarkus-app</input>
                      </fileSet>
                    </fileSets>
                  </comet-aq-webservice>
                </jlink>
              </assemble>
              <release>
                <github>
                  <tagName>{{projectVersion}}</tagName>
                  <releaseName>{{projectVersion}}</releaseName>
                  <overwrite>true</overwrite>
                  <changelog>
                    <formatted>ALWAYS</formatted>
                    <preset>conventional-commits</preset>
                    <format>- {{commitShortHash}} {{commitTitle}}</format>
                    <contributors>
                      <format>- {{contributorName}}</format>
                    </contributors>
                    <labelers>
                      <labeler>
                        <label>task_prep</label>
                        <title>Preparing</title>
                      </labeler>
                      <labeler>
                        <label>task_rel</label>
                        <title>Releasing</title>
                      </labeler>
                    </labelers>
                    <categories>
                      <category>
                        <title>&#x1f9f0; Tasks</title>
                        <labels>task,task_prep,task_rel</labels>
                      </category>
                    </categories>
                    <hide>
                      <categories>&#x1f500; Merge</categories>
                      <contributors>GitHub,dependabot</contributors>
                    </hide>
                  </changelog>
                </github>
              </release>
            </jreleaser>
          </configuration>
        </plugin>
```
 
I think java.xml.bind is the JAXB and it is removed from JDK-11 and hence jdeps is not able to find the required module in JDK -11.

**Reply by aalmiray** [reply]
Could you upgrade to `1.0.0` and try again? Also, do check that all required dependencies have been explicitly copied to an intermediate directory before invoking JReleser's goals like it's done here https://github.com/aalmiray/javafx-canvas-plasma

JReleaser's Maven plugin does not copy JARs based on configured `<dependencies>` at the moment. This step must be invoked explicitly. The linked repo has 3 copy configurations because JavaFX requires platform specific JARs. In your case if you don't have platform specific JARs then a single copy (without listing explicit artifacts) would be sufficient.

**Reply by stackoverflow2** [reply]
Hi @aalmiray  Thanks for the quick responses. I am already using `1.0.0` . As I am using quarkus app, quarkus plugins copy the JAR's based on onfigured `<dependencies>` in the folder `main/target/quarkus-app/lib` and from there using 
` <jars>
                      <jar>
                        <pattern>main/target/quarkus-app/lib/**/*.jar</pattern>
                      </jar>
                    </jars>` 

I am coping the JARs using Jreleaser maven plugin. I can confirm that all JAR's configured using `<dependencies>` are copied to intermediate folder `main\target\jreleaser\assemble\comet-aq-webservice\jlink\inputs\jars\universal`.


Please see trace.log
[trace.log](https://github.com/jreleaser/jreleaser/files/8992113/trace.log)

**Reply by aalmiray** [reply]
I see now. It fails during the `jdeps` step. I have to check if this is a Windows issue (path handling) or generic.

**Reply by stackoverflow2** [reply]
Here are my new findings: 

As explained here[https://blog.bmarwell.de/2020/12/08/use-snakeyaml-in-a-modular-jlink-distribution.html]( https://blog.bmarwell.de/2020/12/08/use-snakeyaml-in-a-modular-jlink-distribution.html)

`Whenever you pull in SnakeYAML (either directly or via Jackson), you will break your modulear builds. The reason: SnakeYAML is a named automatic module. But then, automatic modules cannot be used in jlink images.`

I am using jackson dependency in my project and it has transitive dependency to  `SnakeYAML` . But if this is the case then it should fail in jlink step rather than jdeps step. 

What do you think ?

[trace.log](https://github.com/jreleaser/jreleaser/files/8994454/trace.log)

**Reply by aalmiray** [reply]
Ah auto-modules, yes that's correct, they make jlink fail. However, in this case `jdeps` fails first. 

I've got to think what to make of this situation. For one, JReleaser should allow you to skip `jdeps` explicitly, you must define all input module names as a result.

As an aside, are you looking to package your Quarkus app with jlink as the only packaging option? Is native image an option for you?

**Reply by stackoverflow2** [reply]
Hi @aalmiray we have some constraints to use only jlink packaging option only. We can not use native image.
Why jdeps step is failing in such scenario ? I did not understand. Jdeps list only dependency modules and then jlink will add that all modules in custom JRE. As explained in  [https://blog.bmarwell.de/2020/12/08/use-snakeyaml-in-a-modular-jlink-distribution.html](https://blog.bmarwell.de/2020/12/08/use-snakeyaml-in-a-modular-jlink-distribution.html) we can provide additional modules to jlink.

**Reply by aalmiray** [reply]
Turns out you can disable `jdeps` right now if you supply a set of `moduleNames` to `jlink`. This of course means you have to run jdeps on your own to obtain the set of modules. I've filed #849 to disable `jdeps` explicitly to be consistent.

However, there may be something else happening here as the following [trace.log](https://github.com/jreleaser/jreleaser/suites/7027180421/artifacts/276326224) shows , `jdeps` is able to find the snakeyaml auto-module

```
[DEBUG]   [jlink] D:\a\jreleaser\jreleaser\apps\jreleaser\build\jdks\zulu17Windows\zulu17.32.13-ca-jdk17.0.2-win_x64\bin\jdeps.exe --multi-release base --ignore-missing-deps --print-module-deps universal\artifactory-java-sdk-1.1.0-SNAPSHOT.jar universal\asn-one-0.6.0.jar universal\aws-java-sdk-core-1.12.242.jar universal\aws-java-sdk-kms-1.12.242.jar universal\aws-java-sdk-s3-1.12.242.jar universal\bcpg-jdk15on-1.69.jar universal\bcpkix-jdk15on-1.69.jar universal\bcprov-jdk15on-1.69.jar universal\bcutil-jdk15on-1.69.jar universal\classmate-1.5.1.jar universal\codeberg-java-sdk-1.1.0-SNAPSHOT.jar universal\commons-codec-1.15.jar universal\commons-compress-1.21.jar universal\commons-io-2.11.0.jar universal\commons-lang3-3.12.0.jar universal\commons-net-3.8.0.jar universal\compiler-0.9.10.jar universal\discord-java-sdk-1.1.0-SNAPSHOT.jar universal\eddsa-0.3.0.jar universal\feign-core-11.8.jar universal\feign-form-3.8.0.jar universal\feign-httpclient-11.8.jar universal\feign-jackson-11.8.jar universal\ftp-java-sdk-1.1.0-SNAPSHOT.jar universal\genericgit-java-sdk-1.1.0-SNAPSHOT.jar universal\git-sdk-1.1.0-SNAPSHOT.jar universal\gitea-java-sdk-1.1.0-SNAPSHOT.jar universal\github-api-1.129.jar universal\github-java-sdk-1.1.0-SNAPSHOT.jar universal\gitlab-java-sdk-1.1.0-SNAPSHOT.jar universal\gitter-java-sdk-1.1.0-SNAPSHOT.jar universal\google-chat-java-sdk-1.1.0-SNAPSHOT.jar universal\http-java-sdk-1.1.0-SNAPSHOT.jar universal\httpclient-4.5.13.jar universal\httpcore-4.4.13.jar universal\ion-java-1.0.2.jar universal\jackson-annotations-2.13.3.jar universal\jackson-core-2.13.3.jar universal\jackson-databind-2.13.3.jar universal\jackson-dataformat-cbor-2.13.3.jar universal\jackson-dataformat-toml-2.13.3.jar universal\jackson-dataformat-yaml-2.13.3.jar universal\jakarta.activation-2.0.1.jar universal\jakarta.mail-2.0.1.jar universal\java-sdk-commons-1.1.0-SNAPSHOT.jar universal\JavaEWAH-1.1.12.jar universal\jcl-over-slf4j-1.7.36.jar universal\jmespath-java-1.12.242.jar universal\joda-time-2.8.1.jar universal\jreleaser-1.1.0-SNAPSHOT.jar universal\jreleaser-assemblers-1.1.0-SNAPSHOT.jar universal\jreleaser-config-json-1.1.0-SNAPSHOT.jar universal\jreleaser-config-toml-1.1.0-SNAPSHOT.jar universal\jreleaser-config-yaml-1.1.0-SNAPSHOT.jar universal\jreleaser-engine-1.1.0-SNAPSHOT.jar universal\jreleaser-model-1.1.0-SNAPSHOT.jar universal\jreleaser-templates-1.1.0-SNAPSHOT.jar universal\jreleaser-tools-1.1.0-SNAPSHOT.jar universal\jreleaser-utils-1.1.0-SNAPSHOT.jar universal\jreleaser-workflow-1.1.0-SNAPSHOT.jar universal\jsonschema-generator-4.24.3.jar universal\jsonschema-module-jackson-4.24.3.jar universal\jzlib-1.1.3.jar universal\mail-java-sdk-1.1.0-SNAPSHOT.jar universal\mastodon-java-sdk-1.1.0-SNAPSHOT.jar universal\mattermost-java-sdk-1.1.0-SNAPSHOT.jar universal\org.eclipse.jgit-5.13.0.202109080827-r.jar universal\org.tukaani.xz-0.3.jar universal\os-maven-plugin-1.7.0.jar universal\picocli-4.6.3.jar universal\s3-java-sdk-1.1.0-SNAPSHOT.jar universal\sdkman-java-sdk-1.1.0-SNAPSHOT.jar universal\slack-java-sdk-1.1.0-SNAPSHOT.jar universal\slf4j-api-1.7.36.jar universal\slf4j-simple-1.7.36.jar universal\snakeyaml-1.30.jar universal\ssh-java-sdk-1.1.0-SNAPSHOT.jar universal\sshj-0.33.0.jar universal\teams-java-sdk-1.1.0-SNAPSHOT.jar universal\telegram-java-sdk-1.1.0-SNAPSHOT.jar universal\tika-core-2.4.1.jar universal\tool-sdk-1.1.0-SNAPSHOT.jar universal\twitter-java-sdk-1.1.0-SNAPSHOT.jar universal\twitter4j-core-4.0.7.jar universal\webhooks-java-sdk-1.1.0-SNAPSHOT.jar universal\yamllint-1.5.0.jar universal\zt-exec-1.12.jar universal\zulip-java-sdk-1.1.0-SNAPSHOT.jar
[DEBUG]   [jlink] resolved moduleNames: [java.base, java.desktop, java.management, java.naming, java.rmi, java.security.jgss, java.security.sasl, java.sql]
[DEBUG]   [jlink] moduleNames: [java.base, java.desktop, java.management, java.naming, java.rmi, java.security.jgss, java.security.sasl, java.sql, jdk.crypto.cryptoki, jdk.crypto.ec, jdk.localedata, jdk.net, jdk.security.auth, jdk.security.jgss]
```
